lib/analyzer/image-extractor.ts

```diff
@@ -1,110 +1,181 @@
+export {
+  extractFromTar,
+  SearchActionProducts,
+  SearchActionCallback,
+  mapActionsToFiles,
+  SearchAction,
+};
 
-interface ExtractedKeyFiles {
-  txt: { [key: string]: string };
-  md5: { [key: string]: string };
+const debug = Debug("snyk");
+
+interface SearchActionProducts {
+  [key: string]: { [key: string]: string | Buffer };
 }
 
-interface ExtractedImage {
-  manifest: string;
-  layers: Layers;
+interface SearchActionCallback {
+  name: string; // name, should be unique, for this action
+  callback: (buffer: Buffer) => string | Buffer; // convert Buffer into a string or Buffer
 }
 
-interface Layers {
-  [key: string]: ExtractedKeyFiles;
+interface SearchAction {
+  pattern: string; // path pattern to look for
+  callbacks: SearchActionCallback[]; // array of handlers
 }
 
+/**
+ * Create lookup entries array by mapping the specified callback with each of
+ *  the specified patterns
+ * @param patterns array of path pattern strings
+ * @param callback handler to process files content
+ *  a stream into a string
+ */
+const mapActionsToFiles = (
+  patterns: string[],
+  callback: SearchActionCallback,
+) => {
+  return patterns.map((p) => {
+    return { pattern: p, callbacks: [callback] };
+  });
+};
+
 /**
  * Extract key files form the specified TAR stream.
- * @param stream image layer as a Readable TAR stream
- * @param name layer name
- * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
- * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
- * @returns key files found in the specified layer TAR stream
+ * @param layerTarStream image layer as a Readable TAR stream
+ * @param searchActions array of pattern, callbacks pairs
+ * @returns extracted file products
  */
-async function extractLayerKeyFiles(
-  stream: Readable,
-  txtPatterns: string[],
-  md5Patterns: string[] = [],
-): Promise<ExtractedKeyFiles> {
+async function extractFromLayer(
+  layerTarStream: Readable,
+  searchActions: SearchAction[],
+): Promise<SearchActionProducts> {
   return new Promise((resolve) => {
-    const result: ExtractedKeyFiles = { txt: {}, md5: {} };
+    const result: SearchActionProducts = {};
     const layerExtract: Extract = extract();
-    layerExtract.on("entry", (header, stream, next) => {
-      txtPatterns.forEach((pattern) => {
-        if (minimatch(header.name, pattern, { dot: true })) {
-          streamToString(stream).then((value) => {
-            result.txt[header.name] = value;
-          });
-        }
-      });
-      md5Patterns.forEach((pattern) => {
-        if (minimatch(header.name, pattern, { dot: true })) {
-          streamToBuffer(stream).then((value) => {
-            result.md5[header.name] = md5(value);
-          });
+    layerExtract.on("entry", async (header, stream, next) => {
+      if (
+        header.type === "file" ||
+        header.type === "link" ||
```

Comment:
why do we need the ifs for `link` + `symlink` if you ignore them later?

Reply:
so just split into 2 different ifs - one for `file`and one for the 2 others. This way you'll remove the if inside if